### PR TITLE
[#1903] Add OMS to Offering API

### LINKED ISSUE #1903: Add OMS to offerings API
Expose available OMSes for an offer when creating one through API.

To do that, add an extra endpoint, which will list available OMSes for a resource (service). These will include global OMSes and OMSes registered for the resource's provider.
Each OMS in the response will have at least the following fields: id, custom_offer_details.
Custom_offer_details will be a list of objects, where each object has field: name.

The existing offer POST endpoint will require oms_id and custom_offer_details fields (if offer_type=order_required). Custom_offer_details will be a key-value store.
In case they aren't specified then 400 should be returned with description to go to the new endpoint. (?)

Consider #1964 and https://docs.cyfronet.pl/pages/viewpage.action?spaceKey=FID&title=Studium+przypadku+Oms.custom_params+i+Offer.oms_params.

### PR DESCRIPTION
**Pull request by wujuu**
Closes #1903 

What has changed:
- Add OMS to Offering API offers controller and json schemas
- Add proper_oms validation to offer model
- Add available_oms field to api/v1/resources/ endpoint

How to test locally @Marcelinna:
1. Run this in the terminal:
```
rails db:drop db:create db:migrate && rails dev:prime
```

2. Go to `localhost:5000` and log in
3.  Run this in the terminal:
```
rails c
> User.find_by(email: "email used to log in").update(roles_mask: 11)
```

4. Go to `localhost:5000/backoffice/providers`
5. Find "EGI Federation" and click "edit"
6. Scroll down to "Admins" and add yourself - email used to log in

7. Run this in terminal:
```
rails c
> p1 = Provider.find_by(name: "EGI Federation")
> s1 = Service.find_by(slug: "egi-cloud-compute").update(resource_organisation: p1)
> s2 = Service.find_by(slug: "egi-high-throughput-compute").update(resource_organisation: p1)
> Oms.find_by(default: true).id # note the id
> oms2 = Oms.create(name: "global oms", type: :global) # note the id
> oms3 = Oms.create(name: "provider_group_oms", type: :provider_group, providers: [p1]) # note the id
> oms4 = Oms.create(name: "s1_oms", type: :resource_dedicated, service: Service.find_by(slug: "egi-cloud-compute"), custom_params: {a: {mandatory: true, default: "asd"} }) # note the id
> oms5 = Oms.create(name: "s2_oms", type: :resource_dedicated, service:  Service.find_by(slug: "egi-high-throughput-compute")) # note the id
```

3. Go to `/api_docs` and copy the authentication token
4. Go to `/api_docs/swagger` and add your auth token (**Offering API**)
5. `GET ​/api​/v1​/resources​/{resource_id}`
- resource_id = "egi-cloud-compute" - you should see {"available OMSes": [SOMBO, oms2, oms3 and oms4] (along with other service parameters like `name` and `id`) } and 200 status
- resource_id = "egi-high-throughput-compute" - you should see {"available OMSes": [SOMBO, oms2, oms3 and oms5]  (along with other service parameters like `name` and `id`) } and 200 status

6. `GET ​/api​/v1​/resources​/`
- you should see s1 and s2 with parameters as specified in 5. 

7. `GET  ​/api​/v1​/resources​/{resource_id}​/offers`
- resource_id = "egi-cloud-compute" - you should see some offers and and each should have "primary_oms_id" = SOMBO.id (most likely 1)
- resource_id ="egi-high-throughput-compute" - you should see some offers and each should have "primary_oms_id" = SOMBO.id (most likely 1)

8. `POST  ​/api​/v1​/resources​/egi-cloud-compute/offers`
- payload:
```
{ "name": "experimental offer",
  "description": "sample description",
  "order_type": "order_required",
  "primary_oms_id": oms2.id}
```

Should return 200 and created offer

- payload:
```
{ "name": "experimental offer",
  "description": "sample description",
  "order_type": "order_required",
  "primary_oms_id": oms5.id}
```

Should return 400 and some error mentioning that you are not allowed to use this OMS as primary

- payload:
```
{ "name": "experimental offer",
  "description": "sample description",
  "order_type": "order_required",
  "primary_oms_id": oms4.id}
```

Should return 400 and some error mentioning that oms_params can't be blank

- payload:
```
{ "name": "experimental offer",
  "description": "sample description",
  "order_type": "order_required",
  "primary_oms_id": oms4.id,
  "oms_params": {"asd": "qwe"}
  }
```

Should return 400 and some error mentioning that oms_params have additional unspecified keys

- payload:
```
{ "name": "experimental offer",
  "description": "sample description",
  "order_type": "order_required",
  "primary_oms_id": oms4.id,
  "oms_params": {"a": "yes"}
  }
```

Should return 200 and should return properly created object

- payload:
```
{ "name": "experimental offer",
  "description": "sample description",
  "order_type": "order_required",
  "primary_oms_id": 99999
  }
```

Should return 400 and some error mentioning that OMS does not exist

### CODE REVIEW

File: app/controllers/api/v1/resources/configurations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class Api::V1::Resources::ConfigurationsController < Api::V1::ApiController
+  before_action :find_and_authorize_service, only: :index
+
+  def index
+    render json: { "available OMS": @service.available_oms
```
Comment (jswk) [inline app/controllers/api/v1/resources/configurations_controller.rb:7]: I would stick to the snakecase here too, i.e. `available_omses`.

File: app/models/service.rb [hunk 2]
```diff
@@ -240,6 +240,17 @@ def providers?
       .empty?
   end
 
+  def available_oms
```
Comment (jswk) [inline app/models/service.rb:243]: `available_omses`?

**Comment by jswk** [comment]
Why the decision to merge the trees of offering and resource APIs?

**Comment by wujuu** [comment]
> Why the decision to merge the trees of offering and resource APIs?

/configurations json schema references OMS schema - resulted in swagger error "file not found". Also we have common /v1 API now, but its split into 2 swagger.jsons - In my opinion the only separation needed as the 2 APIs are interconnected. Also the tree is merged only in /swagger directories forming a nice, common schema database

File: app/models/service.rb (hunk 2 above)
Comment (bwilk) [reply app/models/service.rb]: personally I don't like omses :P

File: spec/requests/api/v1/resources/configurations_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  before(:all) do
+    Dir.chdir Rails.root.join("swagger", "v1") # Workaround for rswag bug: https://github.com/rswag/rswag/issues/393
+  end
+
+  after(:all) do
+    Dir.chdir Rails.root # Workaround for rswag bug: https://github.com/rswag/rswag/issues/393
+  end
+
+  path "/api/v1/resources/{resource_id}/configurations" do
+    get "lists projects" do
```
Comment (bwilk) [inline spec/requests/api/v1/resources/configurations_controller_spec.rb:17]: `lists resource configurations ` instead?

**Review by jswk — APPROVED** [review]

**Review by Marcelinna — APPROVED** [review]

**Review by jswk — APPROVED** [review]

**Pull request merged**